### PR TITLE
Support Segmentation Masks in Mixup Layer

### LINKED ISSUE #1550: Support segmentation masks in MixUp layer
This should follow the same structure as segmentation mask augmentation in our other preprocessing layers.

Treating masks like 1-channel images for the purpose of applying MixUp should be a simple and effective approach for this.

### PR DESCRIPTION
**Pull request by cosmo3769**
# What does this PR do?

Fixes #1550 

Here is the [colab](https://colab.research.google.com/drive/1iDq5xaXcaCK369wXPrzWxeTKDJDqbc3U?usp=sharing) link showing the working of segmentation_mask in mix_up layer. 

For the colab code, I took reference from - 

* [CutMix, MixUp, and RandAugment image augmentation with KerasCV](https://keras.io/guides/keras_cv/cut_mix_mix_up_and_rand_augment/)
* [Semantic segmentation with SegFormer and Hugging Face Transformers](https://keras.io/examples/vision/segformer/)

Added an example script to see the result of segmentation mask in mixup layer - 

Screenshot 2023-05-25 at 2 21 47 PM

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? #1550 
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

@ianstenbit @LukeWood

### CODE REVIEW

**Comment by cosmo3769** [comment]
Hey @ianstenbit @LukeWood 

I am also thinking of creating an example of segmentation mask of this layer in `examples/preprocessing/segmentation/`. Should I move forward with creating one in this PR only?

**Comment by ianstenbit** [comment]
> Hey @ianstenbit @LukeWood
> 
> I am also thinking of adding an example of segmentation mask of this layer in `examples/layers/preprocessing/segmentation/` with [Oxford_iiit_pet](https://www.tensorflow.org/datasets/catalog/oxford_iiit_pet) dataset. Should I move forward with creating one in this PR only?

Yes this is a great idea. I think an example would be super helpful

**Review by ianstenbit — CHANGES_REQUESTED** [review]
I think that performing MixUp on non one-hot-encoded segmentation masks is probably not a sound approach, because it could lead to a situation where your label mask includes garbage.

e.g. consider the same pixel in two training frames, one with a mask class of 0, and the other with a mask class of 4.

With this mixup strategy, mixing up those two frames may result in this pixel having a mask class of 1, 2, or 3, which would represent a different class entirely.

Therefore, I think this layer should operate on one-hot-encoded segmentation masks _only_, and indicate this clearly to the user in the docstring (that segmentation masks _must_ be one-hot [or segmentation mask must have only 2 classes) in order for this layer to make sense).

**Comment by cosmo3769** [comment]
Okay, got it. Working on it.

File: keras_cv/layers/preprocessing/mix_up.py [hunk 1]
```diff
@@ -79,6 +80,11 @@ def _batch_augment(self, inputs):
             )
             inputs["bounding_boxes"] = bounding_boxes
         inputs["images"] = images
+        if segmentation_masks is not None:
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/mix_up.py:83]: can you perform input validation on the segmentation masks to ensure they are in one hot encoded format?  This can be done by inspecting the shape, and ensuring it matches [batch, h, w, num_classes].

Comment (cosmo3769) [reply keras_cv/layers/preprocessing/mix_up.py:83]: Yeah, I added input validation for segmentation_mask. Here is the [colab](https://colab.research.google.com/drive/1vz-5MbPgHta9l2FN6c9tgWu1WcyZLrHz?usp=sharing) showing the result.

Also, attaching the image below: 

Screenshot 2023-04-29 at 5 13 22 PM

**Comment by ianstenbit** [comment]
@cosmo3769 are you still working on this / is it ready for another review?

**Comment by cosmo3769** [comment]
Yes @ianstenbit , it is ready for another review. Please check. I will resolve the changes further if required.

File: keras_cv/layers/preprocessing/mix_up_test.py [hunk 2]
```diff
@@ -32,25 +32,35 @@ def test_return_shapes(self):
...
+            outputs["segmentation_masks"],
         )
 
         self.assertEqual(xs.shape, [2, 512, 512, 3])
         self.assertEqual(ys_labels.shape, [2, 10])
         self.assertEqual(ys_bounding_boxes["boxes"].shape, [2, 6, 4])
         self.assertEqual(ys_bounding_boxes["classes"].shape, [2, 6])
+        self.assertEqual(ys_segmentation_masks.shape, [2, 512, 512, 10])
 
     def test_mix_up_call_results(self):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/mix_up_test.py:65]: Let's add a `segmentation_mask` to the test data in this test and make the same assertion (that the resulting masks are, in fact, different)

**Review by ianstenbit — CHANGES_REQUESTED** [review]
This is awesome! 

Sorry for the crazy review latency -- I didn't realize this was ready for another review and it got lost in my inbox.

This looks awesome. I like the colab demo a lot -- how about we add it as an example under `examples/layers/preprocessing/segmentation`? 

We can use the dogs and cats dataset for that I think 😄 

I updated your demo to use the one-hot encoded approach here: https://colab.research.google.com/gist/ianstenbit/a9a5bf0cf9649ff6edb16e38e93845f8/copy-of-segmentaton_mask_mixup.ipynb but it could probably use a little cleanup before we make it an official example

**Comment by cosmo3769** [comment]
> This is awesome!
> 
> Sorry for the crazy review latency -- I didn't realize this was ready for another review and it got lost in my inbox.
> 
> This looks awesome. I like the colab demo a lot -- how about we add it as an example under `examples/layers/preprocessing/segmentation`?
> 
> We can use the dogs and cats dataset for that I think 😄
> 
> I updated your demo to use the one-hot encoded approach here: https://colab.research.google.com/gist/ianstenbit/a9a5bf0cf9649ff6edb16e38e93845f8/copy-of-segmentaton_mask_mixup.ipynb but it could probably use a little cleanup before we make it an official example

Thank you for the review and updating the demo. Should I cleanup the code of this colab and add it as an example or should I use dogs and cats dataset for the examples purpose?

**Comment by ianstenbit** [comment]
> > This is awesome!
> > Sorry for the crazy review latency -- I didn't realize this was ready for another review and it got lost in my inbox.
> > This looks awesome. I like the colab demo a lot -- how about we add it as an example under `examples/layers/preprocessing/segmentation`?
> > We can use the dogs and cats dataset for that I think 😄
> > I updated your demo to use the one-hot encoded approach here: https://colab.research.google.com/gist/ianstenbit/a9a5bf0cf9649ff6edb16e38e93845f8/copy-of-segmentaton_mask_mixup.ipynb but it could probably use a little cleanup before we make it an official example
> 
> Thank you for the review and updating the demo. Should I cleanup the code of this colab and add it as an example or should I use dogs and cats dataset for the examples purpose?

I think using your colab as a base for the example sounds good -- plus it's already using the pets dataset so let's stick with that :)

File: keras_cv/layers/preprocessing/mix_up_test.py (hunk 2 above)
Comment (ianstenbit) [reply keras_cv/layers/preprocessing/mix_up_test.py]: `ys_segmentation_masks = tf.random.uniform(shape=(2, 4, 4), maxval=10)` is going to give you uniform between zero and 1, so the resulting masks after casting to unit8 may not be what you're expecting. Probably we can just use the same `ones` and `2 * ones` approach for the segmentation masks.

Comment (cosmo3769) [reply keras_cv/layers/preprocessing/mix_up_test.py]: Ooh true. Fixed it.

File: examples/layers/preprocessing/segmentation/mix_up_demo.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# Copyright 2023 The KerasCV Authors
```
Comment (ianstenbit) [inline examples/layers/preprocessing/segmentation/mix_up_demo.py:1]: This is great! Could you please add a screenshot of the result of running this script to the PR description?

**Review by ianstenbit — APPROVED** [review]
Looks great! Once you update the description I think we can merge.

Again, apologies for my review latency on this, and thanks for the PR!

File: examples/layers/preprocessing/segmentation/mix_up_demo.py (hunk 3 above)
Comment (cosmo3769) [reply examples/layers/preprocessing/segmentation/mix_up_demo.py:1]: Update. Thank you. @ianstenbit
